src/main/java/com/tabnine/inline/CompletionPreview.java

```diff
@@ -1,250 +1,164 @@
-        && !editor.getSelectionModel().hasSelection()
-        && InplaceRefactoring.getActiveInplaceRenamer(editor) == null) {
-      editor.getDocument().startGuardedBlockChecking();
-
-      try {
-        tabnineInlay.render(this.editor, this.suffix, completion, offset);
-      } finally {
-        editor.getDocument().stopGuardedBlockChecking();
-      }
-      if (!tabnineInlay.isEmpty()) {
-        tabnineInlay.register(this);
-        editor.getContentComponent().addKeyListener(keyListener);
-        editor.addEditorMouseMotionListener(this);
-      }
-    }
-    return suffix;
+    tabnineInlay = TabnineInlay.create(this);
+    caretListener = new InlineCaretListener(this);
+    focusListener = new InlineFocusListener(this);
   }
 
-  public void clear() {
-    if (inApplyMode.get()) {
-      return;
-    }
+  public static TabNineCompletion createInstance(
+      Editor editor, List<TabNineCompletion> completions, int offset) {
+    CompletionPreview preview = getInstance(editor);
 
-    if (!tabnineInlay.isEmpty()) {
-      tabnineInlay.clear();
-      editor.removeEditorMouseMotionListener(this);
+    if (preview != null) {
+      Disposer.dispose(preview);
     }
 
-    editor.getContentComponent().removeKeyListener(keyListener);
+    preview = new CompletionPreview(editor, completions, offset);
 
-    completions = null;
-    suffix = null;
+    editor.putUserData(INLINE_COMPLETION_PREVIEW, preview);
+
+    return preview.createPreview();
   }
 
-  @Override
-  public void dispose() {
-    clear();
-    editor.putUserData(INLINE_COMPLETION_PREVIEW, null);
-    CompletionState.clearCompletionState(editor);
+  @Nullable
+  public static CompletionPreview getInstance(@NotNull Editor editor) {
+    return editor.getUserData(INLINE_COMPLETION_PREVIEW);
   }
 
-  @Override
-  public void mouseMoved(@NotNull EditorMouseEvent e) {
-    alarm.cancelAllRequests();
-    if (tabnineInlay.isEmpty()) {
-      return;
-    }
-    if (e.getArea() == EditorMouseEventArea.EDITING_AREA) {
-      MouseEvent mouseEvent = e.getMouseEvent();
-      Point point = mouseEvent.getPoint();
-
-      if (isOverPreview(point)) {
-        alarm.addRequest(
-            () -> {
-              Point p =
-                  SwingUtilities.convertPoint(
-                      (Component) mouseEvent.getSource(),
-                      point,
-                      editor.getComponent().getRootPane().getLayeredPane());
-              InlineHints.showPreInsertionHint(editor, p);
-            },
-            HINT_DELAY_MS);
-      }
+  public static void clear(@NotNull Editor editor) {
+    CompletionPreview completionPreview = getInstance(editor);
+    if (completionPreview != null) {
+      Disposer.dispose(completionPreview);
     }
   }
 
-  @Nullable
-  public Integer getStartOffset() {
-    return tabnineInlay.getOffset();
+  public void togglePreview(CompletionOrder order) {
+    int nextIndex = currentIndex + (order == CompletionOrder.NEXT ? 1 : -1);
```

Comment:
maybe add a function to the `CompletionOrder` enum so that you can do something like:
```kotlin
int nextIndex = currentIndex + CompletionOrder.NEXT.getProgressionValue();
```

Reply:
Awesome idea!